### PR TITLE
AP_Scripting: move singleton method bindings to flash

### PR DESCRIPTION
**Pull request by IamPete1**
This changes the singleton enum and method tables to be evaluated in `c++`, so they stay in flash and we don't have to load them into memory for lua to use. This method seems to work, no changes required to any scripts. 

This is the memory usage of the simple loop example for master:
```AP: Lua: Time: 0 Mem: 52879 + 72 ```

And with this PR:
```AP: Lua: Time: 0 Mem: 40312 + 616```

So we save 12.5k of memory total but has a bigger increase of +544, I suspect that means I have not got this quite right. ~If we pick up that much on a simple script the memory saving may not be worth it with bigger scripts.~ Seems to not scale and sometimes not pick up anything, I maybe there is some interaction with the garbage clean?

This also costs about 1.6K of flash. 

I have only done singletons here, we could pickup some more savings by also doing userdata and ap object types.

The are also probably better ways to do this if you know what your doing. It would be nice to grab the singleton in the indexing function and then pass it down to the method function, that would save lots of flash, but I don't know how to do that. 

I have not investigated how this effects run times, or the total memory used by the scripting thread.

To get a better idea of what is going on it is easier to compare the generated bindings directly rather than comparing the binding generator.

### CODE REVIEW

**Comment by IamPete1** [comment]
```
Binary           text             data            bss            total
---------------  ---------------  --------------  -------------  ---------------
ardusub          1676 (+0.1113%)  -24 (-1.9108%)  28 (+0.0107%)  1680 (+0.0950%)
blimp            1676 (+0.1358%)  -24 (-1.9802%)  20 (+0.0077%)  1672 (+0.1117%)
arducopter-heli  1976 (+0.1189%)  -24 (-0.9346%)  24 (+0.0092%)  1976 (+0.1027%)
arducopter       1976 (+0.1169%)  -24 (-0.9346%)  24 (+0.0092%)  1976 (+0.1012%)
ardurover        1676 (+0.1111%)  -24 (-1.9544%)  20 (+0.0077%)  1672 (+0.0944%)
antennatracker   1676 (+0.1287%)  -24 (-1.9737%)  20 (+0.0077%)  1672 (+0.1068%)
arduplane        1876 (+0.1108%)  -24 (-0.9693%)  20 (+0.0077%)  1872 (+0.0958%)
```

There is definitely some changes that can be made so this uses less flash, but I think were better to get the lua stuff proven first.

File: libraries/AP_Scripting/generator/src/main.c [hunk 1]
```diff
@@ -1926,23 +1979,28 @@ void emit_metas(struct userdata * data, char * meta_name) {
   fprintf(source, "};\n\n");
 }
 
+void emit_index(struct userdata * data, char * meta_name) {
+  fprintf(source, "const struct userdata_index %s_fun[] = {\n", meta_name);
+  while (data) {
+    start_dependency(source, data->dependency);
+    if (data->enums) {
+      fprintf(source, "    {\"%s\", %s_index, %s_enums},\n", data->alias ? data->alias : data->name, data->name, data->sanatized_name);
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/src/main.c:1987]: we actually don't need the enums here anymore

**Comment by tridge** [comment]
great!
@WickedShell can you review?

**Comment by khancyr** [comment]
@WickedShell this is really interesting for scripting I think, do you have some time to review it ?

**Comment by IamPete1** [comment]
Now implemented the same approach for all singletons, ap_objects and userdata types. Still in memory are the manual bindings and the custom uint32 handling. Simple loop on master uses `55974 + 72` this branch uses `35531 + 616`. So were saving 20K of memory. I think @WickedShell and I decided larger per run memory was a interaction with how the garbage collection works, it is not larger in all cases. 

Looks like were picking up about 1k in flash. Not too bad on the whole. 

I think the main thing now is to test this on lots of different scripts and make sure I have not broken anything and that the runtime memory usage and execution time is not adversely affected. `SCR_DEBUG_OPTS` = 8 for run time logging is very useful for this.

File: libraries/AP_Scripting/generator/src/main.c [hunk 2]
```diff
@@ -2379,6 +2396,43 @@ void emit_docs(struct userdata *node, int is_userdata, int emit_creation) {
 }
 
 
+void emit_index_helpers(void) {
+  fprintf(source, "static bool load_function(lua_State *L, const luaL_Reg *list, const uint8_t length, const char* name) {\n");
```
Comment (tridge) [inline libraries/AP_Scripting/generator/src/main.c:2400]: to get accurate timing on a STM32 need to do the same as ENABLE_EKF_TIMING does
```
#if ENABLE_EKF_TIMING
    void *istate = hal.scheduler->disable_interrupts_save();
    static uint32_t timing_start_us;
    timing_start_us = dal.micros();
#endif
```

**Comment by tridge** [comment]
a benchmark like Tools/CPUInfo would be very useful, along with disabling interrupts

**Comment by IamPete1** [comment]
This test script:
```
param:set("LOG_DISARMED", 1)
param:set("SCR_DEBUG_OPTS", 8)

function update()
  local sensor_count = 0
  for i = 0, 1000 do
    -- num_sensors call is to a singleton that does not take a semaphore
    -- it is at the bottom of the table serach, so should be a worse case test
    sensor_count = sensor_count + rangefinder:num_sensors()
  end

  gcs:send_text(0, "found: " .. tostring(sensor_count))

  return update, 1000
end
return update()
```

Master:

image

This PR:

image

So approx double the run time, `num_sensors` is the 9th item in the list. Vehicle is the largest singleton with 25 items so the last item in that list will be a worse slow down. 

Master memory: `36892 + 368`

This PR:  `22356 + 368`

A saving of `14536` and this is a case where the per call memory is the same.

**Pull request merged**